Sources/MongoSwift/MongoCredential.swift

```diff
@@ -71,5 +71,71 @@ public struct MongoCredential: Decodable, Equatable {
             let string = try container.decode(String.self)
             self.name = string
         }
+
+        internal func getDefaultSource(defaultAuthDB: String?) throws -> String {
+            switch self {
+            case .gssAPI, .mongodbX509:
+                return "$external"
+            case .plain:
+                return defaultAuthDB ?? "$external"
+            case .scramSHA1, .scramSHA256:
+                return defaultAuthDB ?? "admin"
+            default:
+                // The mechanism name has already been validated.
+                throw MongoError.InternalError(message: "Unreachable switch case")
+            }
+        }
+
+        internal func validateCredential(credential: MongoCredential) throws {
+            switch self {
+            case .gssAPI:
+                if credential.username == nil {
+                    throw MongoError.InvalidArgumentError(
+                        message: "Username must be provided for GSSAPI authentication"
+                    )
+                }
+                if let source = credential.source {
```

Comment:
style nit: we can reduce the amount of nesting in some of the checks in this method by combining the `if`s into a single line like
```swift
if let source = credential.source, source != "$external" {
 // ... error
}
```